f/frizzle/comapeo/comapeo_observations.py

```diff
@@ -92,20 +108,50 @@ def fetch_comapeo_projects(
             if project["project_id"] not in comapeo_project_blocklist
         ]
 
-    if not comapeo_projects:
-        raise ValueError("No valid projects found.")
-
     return comapeo_projects
 
 
+@retry(
+    stop_max_attempt_number=3,
+    wait_exponential_multiplier=500,
+)
+def download_attachment(url, headers, save_path):
+    try:
+        response = requests.get(url, headers=headers)
+
+        content_type = response.headers.get("Content-Type", "")
+        extension = mimetypes.guess_extension(content_type) or ""
+
+        file_name = os.path.basename(url) + extension
+        os.makedirs(os.path.dirname(save_path), exist_ok=True)
+        with open(save_path, "wb") as file:
+            file.write(response.content)
+        logger.info("Download completed.")
+        return file_name
+
+    except Exception as e:
+        logger.error(f"Exception during download: {e}")
+        return None
+
+
+def camel_to_snake(name):
+    return re.sub("([a-z0-9])([A-Z])", r"\1_\2", name).lower()
+
+
 def process_comapeo_data(
-    comapeo_server_base_url, comapeo_access_token, comapeo_projects, attachment_root
+    comapeo_server_base_url,
+    comapeo_access_token,
+    comapeo_projects,
+    db_table_prefix,
+    attachment_root,
+    attachment_failed,
 ):
```

Comment:
please add a docstring, see for example the one I added at https://github.com/ConservationMetrics/gc-scripts-hub/blob/9a134d0315c3bb07f69f87e9cd9a8d8e1984e873/f/frizzle/alerts/alerts_gcs.py#L214-L243

More generally, I'd say that a verb like "process" is almost devoid of useful meaning in computer science (all we do is process data, for various definitions of "process" and various examples of "data").  Can this be made more precise?

If not, is it a sign that this function is doing too much?  (NB: reviewing this code, I can also say this function is likely doing too much)

Food for thought; don't sweat all that in this PR.  Just do add the docstring.

Reply:
It seems I failed to mention the preferred docstring format in `CONTRIBUTING.md` ...   I'll comment over on #8.

Reply:
When I first wrote this code for Dagster, I opted to hem closely to the KoboToolbox op graph, which also has a similarly named function `process_forms`, hence this name. But I agree with you.

I'm going to rename it to `download_and_transform_comapeo_data` for now, and will add docstrings to each function.